### ISSUE TITLE
Should colormap/color mode, etc events automatically trigger a refresh?

### ISSUE THREAD

**Issue by jni** [napari/napari#1558]
While @DragaDoncila and I were working on #1555, we noticed that `self.events.color_mode()` did not by itself trigger a refresh of the display. In fact, we needed to subsequently call `self.refresh()`. Is this the intended behaviour? (ie manual refresh in general?) Or should certain classes of events be connected to a refresh automatically?

**Comment by sofroniewn** [comment]
Not intended behaviour, I would not have expected you to need to manually refresh.

**Comment by andy-sweet** [comment]
It's my first time looking at most of this code and I'm a bit confused by the request here, mostly because there seem to be some important differences between the two properties mentioned (colormap and color_mode).

The `IntensityVisualizationMixin::colormap` directly corresponds to the `VispyImageVisual::cmap` property, so refreshing the display is mostly as simple as keeping those properties in sync. This is done by connecting `VispyImageLayer::_on_colormap_change` to the `colormap` event, which is emitted at the end of the `colormap` setter.

The `Labels::color_mode` property doesn't directly correspond to a vispy property. Instead, when updating its value, we need to do some fairly complex updates to other state of the same Labels object. Most importantly `Layer::slice::image::_view` needs to be reconverted based on the new colormap indexing that corresponds to the new color_mode value. `Layers::refresh` handles these updates, and then emits the `Layers::set_data` event, which is connected `VispyBaseLayer::_on_data_changed` which handles refreshing the display (now that the state of the Labels Layer is good).

Is the request here to connect a method of `VispyImageLayer` to the `color_mode` event? Wouldn't that just call `Layers::refresh` anyway? 

Or is the request here to do something less blunt than calling `Layers::refresh`? I think that's possible, but I'm still not sure that adding an extra connection to `VispyImageLayer` will help.

@jni - any clarification would be much appreciated!

**Comment by jni** [comment]
The request is to refresh the layer after setting color_mode, bluntly or not. 😂 Preferably not, but bluntly (ie connecting the color_mode setter to layer.refresh) works as a first try!

**Comment by andy-sweet** [comment]
At the risk of being annoying/pedantic, we already refresh the layer after setting its color_mode - but through an explicit call rather than through a connection to a method of VispyImageLayer. I think the explicit call has a couple of advantages over a new connection.

1. Simpler program flow and less indirection (Layer::refresh was already pretty tricky to follow!).
2. The call to Layer::refresh is updating some state of Layer that is effectively invalid due to the change in color_mode. It seems more natural for Layer to handle that itself, rather than rely on VispyImageLayer. Plus, if any other object depends on that state in the future, then it's better to update it once in Layer before any other object is notified about the color_mode event. 

That being said, if we do something less blunt than refresh (which itself emits an event that VispyImageLayer is connected to), then I think it could make sense to add a connection in VispyImageLayer, but still only after Layer has cleaned its own state up.

My guess is that I'm missing something still, which might be resolved with sleep, but feel free to correct the logic above! It might also make sense to try to quickly resolve this in the dev sync on Wednesday, rather than through comments here.

**Comment by sofroniewn** [comment]
it could be that when this issue was raised back last year we were missing that explicit call that @andy-sweet notes. Can you check @jni if the behavior is now fixed? There have been some big changes to this part of the code base

**Comment by sofroniewn** [comment]
> In fact, we needed to subsequently call self.refresh()

i.e. do you as a user of napari still need to explicitly make this call @jni? if I read @andy-sweet correctly then you the user don't need to make this call anymore. A minimal example here might also really help clear this up

**Comment by andy-sweet** [comment]
Based on the initial description, it sounded like there was an expectation that calling `Labels::events::color_mode()` would cause the refresh. That's still not true, as you need to call the `color_mode` setter. Agreed an example would help to clarify that.

I think I also made an assumption that the events (while publicly exposed) represent more of an implementation detail, and that using the setter is the intended way for clients to interact with a `Labels` layer. So any guidance there would be useful too.

**Comment by sofroniewn** [comment]
If I run `python examples/add_labels_with_properties.py` and in the console do `viewer.layers[1].color_mode = 'auto'` then the display updates automatically without needing to call `layer.refresh`. That is the behavior that matters to me, and I think when I said https://github.com/napari/napari/issues/1558#issuecomment-674934908 I might have misunderstood thinking that was not the case, not really appreciating the issue was about `self.events.color_mode()` itself doesn't trigger the update.

Given that is what the issue is about, I want to take back by comment at the top saying that i would have expected that to work - I actually don't think that's the case, and I don't really see users manually emitting events themselves (though I guess they can using our public API).

My hope here is that once we start the pydantic refactor for the layers this will actually force some clean up in some of these slightly circuitous places like we have right now, so I'd be tempted to close this issue

**Comment by jni** [comment]
Yeah I don't fully remember the context. If setting color_mode refreshes, I think that's good enough. Could well be that this issue can be closed. I'll reopen if I run into this again! Sorry for the confusion!

**Comment by jni** [comment]
Actually, maybe it should stay open. 😅 Specifically: rather than the color_mode setter calling `self.refresh()`, shouldn't we *only* call `self.events.color_mode()` and trigger a refresh from that? ie connect events that affect the display to `self.refresh`?

**Comment by sofroniewn** [comment]
Calling refresh inside these setters is our standard pattern right now - for example see `layer.mode` setter. 

https://github.com/napari/napari/blob/a279594e677f913067ce65c8fc62943d1819900c/napari/layers/labels/labels.py#L724-L725

I think we should just defer this until the pydantic refactor which will force us to standardize event emission on our layers

**Comment by jni** [comment]
To summarise a conversation @sofroniewn and I just had offline:

- Currently, all our layers use `self.refresh` to trigger a re-draw when something changes that requires such a thing. This is intended, a one stop shop for redrawing.
- However, that might actually be doing more work than necessary: `refresh` triggers re-slicing, which could be expensive and unnecessary work when e.g. changing the color mode.
- It *might* be desirable in the future to be more surgical with our events: `color_mode` triggers an event and the vispy view/other views hook into that, and only update the elements of the view that need to change.
- However, this is in no way urgent and probably fits within a much bigger discussion, e.g. #1353. (For example, thumbnails should be a view, not part of the model.)

So, the initial classification of this issue as priority-high was inappropriate, and stemmed from an initial idea that `layer.color_mode = foo` was not updating, which was not the case — @sofroniewn has updated above. 

I've advocated that this should stay in the ice box but open, as we don't yet have a more general issue to track how our smaller events should work.

**Comment by andy-sweet** [comment]
That summarizes my thoughts too, but more concisely than I was able to manage! If I have some time in the short term, I might try to make `color_mode` behave more surgically and see where that goes, but will probably run away if it looks like much bigger changes are needed.
